### ISSUE TITLE
about code 

### ISSUE THREAD

**Issue by kwk12306** [ahmdtaha/knowledge_evolution#7]
Hi Ahmed! 
      Your paper helps me a lot!  I have doubts  about these，
①the function of  variable  “in_channels_order” such as in the function  conv3x3(self, in_planes, out_planes, stride=1, first_layer=False,bias=False,in_channels_order=None)
②class BasicBlock(nn.Module):M = 2  ，the function  of  variable  M
Thanks！

**Comment by ahmdtaha** [comment]
Hi LeoLiu2n,

1. `in_channels_order` has no value for ResNets, but vital for [DenseNets](https://github.com/ahmdtaha/knowledge_evolution/blob/a3f2eb2eed7accb86ad1af2a15c13e4a9654fe16/models/split_densenet.py#L42). The [paper Appendix.B](https://arxiv.org/pdf/2103.05152.pdf) illustrates the importance of this variable (Fig. 10 and 11)

2. As mentioned in the readme file, I borrowed the ResNet architecture from [Vivek Ramanujan and Mitchell Wortsman's repos](https://github.com/allenai/hidden-networks). They had this `M` in [their code](https://github.com/allenai/hidden-networks/blob/dddf2d093de568fc76d460a77fa2650e56e79c1a/models/resnet.py#L8), I never removed it.

I hope this helps

**Comment by kwk12306** [comment]
Hi Ahmed!
       Thank you for patiently explained! but I meet some new doubts.
        1. arg_reset_hypothesis = False，its help description says  "Reset hypothesis across generations"，does it mean train H▽ at the same time？
        2. I saw  such as  Epoch: Data  0.024 ( 0.091)  Loss 1.044 (0.892)	Acc@1  92.86 ( 99.31)  in the train logs ，0.892 means val loss？99.31 means  val acc TOP1？
        3. can kels  apply with pruning methods such as filter pruning ？as the paper say  KELS belongs to the filter-
pruning category. or apply with classic knowledge distillation？because  I  saw it combined with CS-KD.
        4. This long training time can be re-duced by monitoring the performance on a validation split, does it mean I  can  add
  early stopping  in every generation ?
        looking  forward to hearing from you again！

**Comment by ahmdtaha** [comment]
Hi LeoLiu2n,

1. In vanilla KE, the fit and reset hypotheses' binary-masks (Fig. 3) are fixed once initialized. Yet, it doesn't have to be like that. E.g., in the [Appendix-D: Ablation Study (Fig. 16)](https://arxiv.org/pdf/2103.05152.pdf), I randomly re-initialized the binary mask -- that outlines the fit-hypothesis -- at the end of each generation. To do so, I set ` arg_reset_hypothesis=True`. This never boosted performance, but it was a nice ablation study. 

2. Loss `x` (`y`) means the _current_ mini-batch loss = `x`, while the average loss so far = `y` as shown [here](https://github.com/ahmdtaha/knowledge_evolution/blob/a3f2eb2eed7accb86ad1af2a15c13e4a9654fe16/utils/logging.py#L86). Basically, you should track and use `y`. `x` is a transient snapshot.

3. I don't understand the question. KELS delivers a slim network. Thus, KE+KELS can be regarded as a pruning approach independent of the loss-function (with or without knowledge distillation).

4. You can apply early stopping at every generation. However, this is not what I meant in the `Discussion` paragraph. What I meant is that there is no way to automatically determine the ideal number of generations per dataset. E.g., in Fig. 1 I used 100 generations which is big. One way to decide whether to terminate training -- stop at a certain `g` or not -- is to monitor the performance on the validation setup.

I hope this helps

**Comment by kwk12306** [comment]
Hi Ahmed!
     I used cmd python train_KE_cls.py --resume './data/checkpoints/KE/SPLT_CLS_Aircraft100Pytorch_Split_ResNet18_CS_KDTrue_smthFalse_k0.5_G10_e200_evrand_hResetFalse_smkels_single_gpu_test3/gen_0/split_rate=0.5/0002_g0/checkpoints/model_best.pth'  ,but it meets error"join() argument must be str, bytes, or os.PathLike object, not 'NoneType'"so could you tell me correct way for checkpoint running? Thanks a lot!!!

**Comment by ahmdtaha** [comment]
Hi LeoLiu2n

I couldn't replicate your error. I tried the `resume` option myself and seems to work. My current guess is that you forget to **explicitly** provide other arguments.  You can't simply use `resume` alone. For example, you must provide an experiment name to avoid crashing at [this line](https://github.com/ahmdtaha/knowledge_evolution/blob/a3f2eb2eed7accb86ad1af2a15c13e4a9654fe16/configs/base_config.py#L288). 

Whenever reporting similar errors, please report both the module and line # causing the error. E.g., you said "join() argument ...", but there are many `join` in the repos!

Thanks

**Comment by kwk12306** [comment]
Hi Ahmed!
    Thanks for your patiencea again! With your help, I has solved the resume problem. Also, I want to know which should I  load?   model_best.pth or epoch_199.pth ?(they are all saved in the last generation's checkpoints flie )   when  I use "resume"because the unknowed physcis interrupt  in the process of   a  generation training. Thanks!

**Comment by ahmdtaha** [comment]
The model_best.pth is the checkpoint that achieved the best performance on the validation set, while epoch_199.pth is the last checkpoint. I consistently use the last epoch. I know other people who prefer the checkpoint with best performance on the validation set. It is a personal choice
